tmt/steps/__init__.py

```diff
@@ -324,44 +387,58 @@ def __init__(self, step, data, workdir=None):
             name=data['name'],
             workdir=workdir,
             order=order)
-        self.data = data
+        # It is not possible to use TypedDict here because
+        # all keys are not known at the time of the class definition
+        self.data: Dict[Any, Any] = data
         self.step = step
 
     @classmethod
-    def options(cls, how=None):
+    def base_command(cls, method_class: Optional[Method] = None,
+                     usage: Optional[str] = None) -> click.Command:
+        """ Create base click command (common for all discover plugins) """
+        raise NotImplementedError
+
+    @classmethod
+    def options(cls, how: Optional[str] = None) -> List[click.Option]:
         """ Prepare command line options for given method """
         # Include common options supported across all plugins
+        assert isinstance(tmt.options.verbose_debug_quiet, list)
+        assert isinstance(tmt.options.force_dry, list)
         return tmt.options.verbose_debug_quiet + tmt.options.force_dry
 
     @classmethod
-    def command(cls):
+    def command(cls) -> click.Command:
         """ Prepare click command for all supported methods """
         # Create one command for each supported method
-        commands = {}
-        method_overview = f'Supported methods ({cls.how} by default):\n\n\b'
+        commands: Dict[str, click.Command] = {}
+        method_overview: str = f'Supported methods ({cls.how} by default):\n\n\b'
         for method in cls.methods():
             method_overview += f'\n{method.describe()}'
-            command = cls.base_command(usage=method.usage())
+            command: click.Command = cls.base_command(usage=method.usage())
             # Apply plugin specific options
             for option in method.class_.options(method.name):
                 command = option(command)
```

Comment:
Shouldn't it be `command = command(option)` instead?

Reply:
I believe these are correct.

`class_.options` is a list of `click.option` instances, e.g.

```python
force_dry = [
    click.option(
        '-f', '--force', is_flag=True,
        help='Overwrite existing files and step data.'),
    click.option(
        '-n', '--dry', is_flag=True,
        help='Run in dry mode. No changes, please.'),
    ]
```

Normally, one would apply these as decorators:

```python
@click.option('-f', '--force', ...)
@click.option('-n', '--dry', ...)
def some_command(...):
```

`@` is a lump of syntax sugar, what's happening automagically is this:

```python
def some_command(...):
    ...

some_command = click.option('-f', ...)(click.option('-n', '--dry', ...)(some_command))
```

So the code does this, in a slightly unexpected fashion. Each option is "added" to a given command, which means "calling" `option` with `command` as its parameter. Functions that are used as decorators are then expected to return new callable, with whatever modifications they needed to add, and this value is assigned to the original name. Hence the loop over all options.

Reply:
I see, thank you. Ok, I suppressed the error here.

Reply:
What errors was it, btw?

Reply:
It was `error: "Option" not callable`